### PR TITLE
feat(target): Add "docker" target

### PR DESCRIPTION
**Pull request by BYK**
This change adds a "docker" target which pulls an already built
source image tagged with the revision being released, tags it with
the new version, and pushes to a new target.

### CODE REVIEW

**Comment by BYK** [comment]
Not adding tests as there are no tests for most of the targets and the test would essentially replicate the implementation (as there are no edge cases to cover, or even too much complexity)

File: src/targets/docker.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+/**
+ * Command to launch docker
+ */
+const DOCKER_BIN = process.env.DOCKER_BIN || DEFAULT_DOCKER_BIN;
+
+/** Options for "docker" target */
+export interface DockerTargetOptions extends TargetConfig {
+  username: string;
+  password: string;
+  /** Source image path, like `us.gcr.io/sentryio/carft` */
```
Comment (tonyo) [inline src/targets/docker.ts:21]:
```suggestion
  /** Source image path, like `us.gcr.io/sentryio/craft` */
```
🙈

File: src/targets/docker.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    return spawnProcess(DOCKER_BIN, ['push', target], {}, { showStdout: true });
+  }
+
+  /**
+   * Pushes a source image to Docker Hub
+   *
+   * @param version The new version
+   * @param revision The SHA revision of the new version
+   */
+  public async publish(_version: string, revision: string): Promise<any> {
```
Comment (tonyo) [inline src/targets/docker.ts:114]:
```suggestion
  public async publish(version: string, revision: string): Promise<any> {
```

File: src/targets/docker.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    await this.login();
+    await this.pull(revision);
+    await this.push(revision, _version);
```
Comment (tonyo) [inline src/targets/docker.ts:117]:
```suggestion
    await this.push(revision, version);
```

File: src/targets/docker.ts (hunk 2 above)
Comment (tonyo) [reply src/targets/docker.ts:114]: `_version` was there just when it was unused, IIRC.

Comment (BYK) [reply src/targets/docker.ts:114]: Yup, oversight. Thanks for the heads up :)

**Comment by BYK** [comment]
>Probably stating the obvious, but for this target to work, the system that builds Docker images (GCB, Docker Hub, Travis, etc.) should report its status to GitHub as a commit status check (or whatever it's called now), so that craft publish (with GitHub as status provider configured) can pick it up and wait until the image is ready.

I mean, that would be _ideal_ but if the image doesn't exist, I just expect Craft to fail as `docker pull` will error out. Makes sense?

**Comment by tonyo** [comment]
> I mean, that would be ideal but if the image doesn't exist, I just expect Craft to fail as docker pull will error out. Makes sense?

Yes, it will fail, it's just not all the targets in craft are currently idempotent, meaning that if the publishing process fails on this target, it'll require some human involvement to apply the remaining targets. 
So ideally, one should make sure that all required artifacts (be it files or docker images) are present before running `publish`. For files, craft can do that, but in this case there's no artifact we upload, so the idea is to check that Docker build finished successfully.

File: src/targets/docker.ts (hunk 1 above)
Comment (jan-auer) [reply src/targets/docker.ts]: Random spam: Let's rename this repo to carft. I really like the sound of it. So quirky :)

File: README.md [hunk 4]
```diff
@@ -776,6 +776,48 @@ targets:
     specPath: MyProject.podspec
 ```
 
+### Docker (`docker`)
+
+Pulls an existing source image tagged with the revision SHA, and then pushed it
+to a new target tagged with the released version. No release
+assets are required for this target except for the source image at the provided
+source image location so it would be a good idea to add a status chcek that
```
Comment (tonyo) [inline README.md:784]:
```suggestion
source image location so it would be a good idea to add a status check that
```

File: README.md [hunk 5]
```diff
@@ -776,6 +776,48 @@ targets:
...
+
+**Environment**
+
+`docker` executable (or something equivalent) must be installed on the system.
+
+| Name                    | Description                               |
+| ----------------------- | ----------------------------------------- |
+| `DOCKER_USERNAME`       | The username for the Docker registry      |
+| `DOCKER_PASSWORD`       | The password for the Docker registry      |
+| `DOCKER_BIN`            | **optional**. Path to `pod` executable.   |
```
Comment (tonyo) [inline README.md:797]:
```suggestion
| `DOCKER_BIN`            | **optional**. Path to `docker` executable.   |
```

File: README.md [hunk 6]
```diff
@@ -776,6 +776,48 @@ targets:
...
+| Option     | Description                                  |
+| ---------- | -------------------------------------------- |
+| `source`   | Path to the source Docker image to be pulled |
+| `target`   | Path to the target Docker image to be pushed |
+
+**Example**
+
+```yaml
+targets:
+  - name: cocoapods
```
Comment (tonyo) [inline README.md:810]:
```suggestion
  - name: docker
```
